chore(rust): add ockam status command

hargut opened the pull request:
## Current Behavior
No `status` command available.

## Proposed Changes
Introduce `status` command, providing information about enrolled or all active identities and linked nodes.

Solves https://github.com/build-trust/ockam/issues/3460

## Checks
- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [x] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository.

<!-- Looking forward to merging your contribution!! -->

hargut commented:
Example output from  `ockam status`:
```
Identity[0]
  Default: yes
  Name: da80759c
  State Path: /home/hargut/.ockam/identities/da80759c.json
  Config Identifier: P36cae565dfde8d83f83d89d63d91a175ede4f6ea6bf799fc4ccfe8a4c8aa29b3
  Linked Nodes:
    Node[0]:
      Name: 0afffc7d
      Status: Stopped
```

adrianbenavides commented:
Thanks @hargut for submitting this PR. From what I understood reading the [issue](https://github.com/build-trust/ockam/issues/3460), this command should show the identities that were enrolled with `ockam enroll`, and not just listing the available identities.

To do that, you should modify the `IdentityConfig` (the information we store about identities on disk) with a new field to track whether an identity was enrolled. This information should be filled after an identity has been successfully enrolled (`enroll` command). Then, on the `IdentitiesState`, you could add a method to retrieve the identities that were enrolled. For now, you can keep the shown information simple (i.e. no need to show the linked nodes). 

Let me know if you have any questions or need help with anything!

Commit:
chore(rust): add ockam status command

hargut commented:
Hi @adrianbenavides,

I've added the part to store the enrollment status with the identity, and modified the `status` command to only display the information for enrolled identities. With the `--all` flag details will be printed for all identities including non-enrolled.
```
Identity[0]
  Default: yes
  Name: f6946397
  State Path: /home/hargut/.ockam/identities/f6946397.json
  Config Identifier: P47018045bdfeeb1c55158d7725ebdda8cabb8374efc4db3c3b08a791e77113b4
  Enrollment Status:
    Enrolled: yes
    Timestamp: 1671975757
  Linked Nodes:
    Node[0]:
      Name: 9f82ddc9
      Status: Running
```

adrianbenavides commented on implementations/rust/ockam/ockam_api/src/cli_state.rs:
```diff
@@ -322,6 +373,29 @@ impl PartialEq for IdentityConfig {
+        }
+
+        match self.created_at.duration_since(UNIX_EPOCH) {
+            Ok(time) => writeln!(f, "Timestamp: {}", time.as_secs())?,
```
I think it would be best to show a human-readable format here. I'd suggest adding the [`time`](https://github.com/time-rs/time) dependency and format the `SystemTime` to [Iso8601](https://docs.rs/time/latest/time/format_description/well_known/struct.Iso8601.html#)

hargut replied:
I've upfront already considered and tested that approach, but it lead to an issue within the no_std check in the rust workflow. Therefore I've for now just included the POSIX style output. So far I've not tested the chrono crate which might be another solution.

https://github.com/hargut/ockam/blob/feature/add-status-command/.github/workflows/rust.yml#L159
https://github.com/hargut/ockam/actions/runs/3776101474/jobs/6419173662#step:7:65

https://github.com/chronotope/chrono/blob/378efb157d674c01761f538d4450705c2b1766a4/.github/workflows/test.yml#L87

hargut replied:
@adrianbenavides
Please let me know your thoughts on how to handle that point.

adrianbenavides replied:
OK, let's not worry about this. I'll open a new issue to tackle this in a separate PR.

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/enroll.rs:
```diff
@@ -357,3 +360,27 @@ impl Auth0Service {
+    node_name: &str,
+) -> Result<()> {
+    let mut identities = opts.state.identities.list()?;
+    identities.push(opts.state.identities.default()?);
```
Why is this needed? Shouldn't the default identity should be returned by `list()`?

hargut replied:
Thank you for pointing that out, I've checked it again and it is not needed.

The locations where two different directories so I had assumed it is two different files, but the default identity is just a symlink to the identities directory.

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/status.rs:
```diff
@@ -0,0 +1,125 @@
+            println!("{:2}Linked Nodes:", "");
+            for (n_idx, node) in node_details.iter().enumerate() {
+                println!("{:4}Node[{}]:", "", n_idx);
+                if node.identity.identifier() == &identity.config.identifier {
```
Note that the two prints done before this line will be printed for every identity, even if only one identity has linked nodes. You should buffer the print contents and print it after you know for sure that the current identity has linked nodes.

hargut replied:
Good point, I'll adjust that part.

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/enroll.rs:
```diff
@@ -357,3 +360,27 @@ impl Auth0Service {
+
+    for mut identity in identities {
+        if node_identity.identifier() == &identity.config.identifier {
+            identity.config.enrollment_status = Some(EnrollmentStatus {
+                is_enrolled: true,
+                created_at: SystemTime::now(),
+            });
```
Since the `EnrollmentStatus` variables are constants, you could add a constructor so you can write this line like `identity.config.enrollment_status = Some(EnrollmentStatus::enrolled());`.

hargut replied:
Will adjust that. :+1:

adrianbenavides reviewed (changes requested):
Looks pretty great! I've left some suggestions, thanks! 👍

hargut commented:
Hi @adrianbenavides,

thank you for your review. Happy New Year!

Best regards,
Harald

adrianbenavides reviewed (dismissed):
Thanks again @hargut and happy new year!
LGTM 🚀
